include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,188 @@
+#include <eve/memory/aligned_allocator.hpp>
+#include <eve/concept/simd_allocator.hpp>
+#include <eve/product_type.hpp>
+#include <type_traits>
+#include <memory>
+
+namespace eve::algo::detail
+{
+  template<eve::product_type Type, eve::simd_allocator Allocator>
+  struct soa_storage : Allocator
+  {
+    using is_product_type = void;
+
+    soa_storage() noexcept
+                : indexes_{}, storage_( nullptr, aligned_deleter{this} ), size_{}, capacity_{}
+    {}
+
+    explicit  soa_storage( std::size_t n) : soa_storage( n, Type{} ) {}
+
+    soa_storage( std::size_t n, Type const& v)  : soa_storage()
+    {
+      storage_t  local{ reinterpret_cast<byte_t*>(Allocator::aligned_alloc(byte_size(n), 64))
+                      , aligned_deleter{this}
+                      };
+      storage_.swap(local);
+      size_     = n;
+      build(n,v);
+    }
+
+    soa_storage(soa_storage const& src) : soa_storage()
+    {
+      auto sz = byte_size(src.size_);
+      storage_t  local{ reinterpret_cast<byte_t*>(Allocator::aligned_alloc(sz,64))
+                      , aligned_deleter{this}
+                      };
+      std::copy(src.storage_.get(), src.storage_.get()+sz, local.get());
+
+      storage_.swap(local);
+      indexes_  = src.indexes_;
+      size_     = src.size_;
+      capacity_ = src.capacity_;
+    }
+
+    soa_storage& operator=(soa_storage const& src)
+    {
+      *this = soa_storage(src);
+      return *this;
+    }
+
+    soa_storage(soa_storage&&)            =default;
+    soa_storage& operator=(soa_storage&&) =default;
+
+    EVE_FORCEINLINE auto data_aligned()
+    {
+      auto ptrs = kumi::map([]( auto p) { return as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([]( auto p) { return as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      std::swap(indexes_,other.indexes_);
+      storage_.swap(other.storage_);
+      std::swap(size_,other.size_);
+      std::swap(capacity_,other.capacity_);
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept
+    {
+      a.swap(b);
+    }
+
+    void build(std::size_t n, Type v = {})
+    {
+      auto sub_span = storage_.get();
+      auto offset = 0;
+
+      kumi::for_each_index( [&]<typename Idx>(Idx, auto& s, auto m)
+                            {
+                              // Constructs the current sub-span
+                              using type = kumi::element_t<Idx::value,Type>;
+                              auto p = reinterpret_cast<type*>(sub_span+offset);
+                              for(std::size_t i=0;i<n;++i) new (p+i) type(m);
```

Comment:
This is bad. Because this will generate code. We want to just construct with garbage and then use eve algo to fill in the values

Reply:
Ok so offset computation then fill